src/contexts/Auth.tsx

```diff
@@ -0,0 +1,65 @@
+import { ComponentType, createContext, useContext } from "react";
+import { Navigate, useLocation } from "react-router-dom";
+import Icon from "components/Icon";
+import LoaderRing from "components/LoaderRing";
+import { useGetter } from "store/accessors";
+import { AuthenticatedUser, PreRenderCredential } from "slices/auth";
+import { appRoutes } from "constants/routes";
+
+export default function withAuth<Props>(
+  Component: ComponentType<Props & { user: AuthenticatedUser }>,
+  requiredCredentials?: PreRenderCredential[]
+) {
+  return function Protected(props: Props) {
+    const location = useLocation();
+    const user = useGetter((state) => state.auth.user);
+
+    if (user === "loading" || user?.isSigningOut) {
+      return (
+        <div className="grid content-start place-items-center py-8">
+          <LoaderRing thickness={12} classes="w-32 mt-8" />
+        </div>
+      );
+    }
+
+    if (!user) {
+      return (
+        <Navigate to={appRoutes.signin} state={{ from: location }} replace />
+      );
+    }
+
+    if (
+      !(requiredCredentials || []).every((c) => user.credentials.includes(c))
+    ) {
+      return (
+        <div className="grid content-start place-items-center py-20">
+          <Icon type="ExclamationCircleFill" size={80} className="text-red" />
+          <p className="text-xl mt-8 font-work ">Unauthorized</p>
+        </div>
+      );
+    }
+
+    /**
+     * if wish to access user via props wrap component:
+     * withAuth(function Component({user, ...restProps}){
+     * })
+     */
+    return (
+      <Context.Provider value={user}>
+        <Component {...props} user={user} />
+      </Context.Provider>
+    );
+  };
+}
+
+const INIT = "__init__" as unknown as AuthenticatedUser;
+const Context = createContext<AuthenticatedUser>(INIT);
```

Comment:
Just to make sure I understood this code, the reason for assigning `INIT` here it so that `AuthenticatedUser` is always defined?

Reply:
> so that AuthenticatedUser is always defined?
another reason is to have a way to distinguish init context value as `useAuthenticatedUser` would result to `INIT` if used outside `withAuth`

we could set this to just `undefined` but would also need to coerce to silence typescript error
```ts
const Context = createContext<AuthenticatedUser>() //ts-error
const Context = createContext<AuthenticatedUser>(undefined as unknown as AuthenticatedUser) 
```
and check in hook as 
```js
//useAuthenticatedUser

if val = useContext(context)
if(val === undefined) // error: used outside withAuth
```

we can't however type like this 
```ts
const Context = createContext<AuthenticatedUser | undefined>() 
```
as consumers would then receive
```
AuthenticatedUser | undefined 
```
which is a case that isn't a thing but still needs to be checked everytime the hook is used

Reply:
@ap-justin 
> we could set this to just undefined but would also need to coerce to silence typescript error
> ```ts
> const Context = createContext<AuthenticatedUser>() //ts-error
> const Context = createContext<AuthenticatedUser>(undefined as unknown as AuthenticatedUser)
> ```

That's of course necessary and the same exact thing is done for `INIT`
```ts
const INIT = "__init__" as unknown as AuthenticatedUser;
const Context = createContext<AuthenticatedUser>(INIT);
```

I understand what you were trying to achieve now, doesn't really matter which default value we use